cryostatio/cryostat
# fix(rules): refactor rules to become aware of jvmIds

# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
_______________________________________________

Fixes: #1092 

## Description of the change:
Rules are not aware of jvmIds and will do operations more than once on targets if there exist more than one jvmId unique target within the discovery storage. This PR allows rules to do only do operations on jvmId unique targets with non-null jvmIds. This should work since the discovery storage is updated with non-null jvmIds whenever credentials are posted correctly and targets with the same jvmId are the same jvm. 

However there is one problem in that when Cryostat restarts, the custom targets that are persisted never update their own jvmId when Cryostat comes back up, and so there will be a mismatch of jvmIds and the custom target will have a "stale" id. I'm not sure what is the best way to deal with this. Maybe on built-in discovery startup, test target connections for their jvmId again? I'm not too sure about the race conditions for that problem for now.

## Motivation for the change:
JvmIds are unique to jvms. We don't want rule operations, for example deleting/stopping a recording, to apply more than once on a target jvm as that is extra overhead cost.

## How to manually test:
1. *Run CRYOSTAT_IMAGE=quay.io... sh smoketest.sh...*
2. Create a target "cryostat:9093"
3. Create a rule with matchExpression=="true"
4. Enable the rule. 
5. Disable the rule or delete the rule. The logger should no longer show an error.
